iavl/iavl_proof2.go

```diff
@@ -167,52 +171,266 @@ func (proof *KeyNotExistsProof) Verify(key []byte, root []byte) error {
+	}
+	inner := fmt.Sprintf("PathToKeys: \n\t%s", strings.Join(paths, ",\n"))
+	inner += fmt.Sprintf("LeftNode: %#v (%x)\n", proof.LeftNode, proof.LeftNode.Hash())
+	inner += fmt.Sprintf("RightNode: %#v (%x)", proof.RightNode, proof.RightNode.Hash())
+	return "&KeyRangeProof{\n\t" + inner + "\n}"
 }
 
-func (proof *KeyRangeExistsProof) Verify(key []byte, value []byte, root []byte) bool {
-	return false
+// Returns a list of all paths, in order.
+func (proof *KeyRangeProof) paths() []*PathToKey {
+	paths := proof.PathToKeys[:]
+	if proof.LeftPath != nil {
+		paths = append([]*PathToKey{proof.LeftPath}, paths...)
+	}
+	if proof.RightPath != nil {
+		paths = append(paths, proof.RightPath)
+	}
+	return paths
 }
 
-func (node *IAVLNode) constructKeyExistsProof(t *IAVLTree, key []byte, proof *KeyExistsProof) ([]byte, error) {
+// Checks that all paths are adjacent to one another, with no gaps.
+func (proof *KeyRangeProof) verifyPathAdjacency() error {
+	paths := proof.paths()
+	for i := 0; i < len(paths)-1; i++ {
+		// Always check from left to right, since paths are always in ascending order.
+		if !paths[i].isLeftAdjacentTo(paths[i+1]) {
+			return errors.Errorf("paths #%d and #%d are not adjacent", i, i+1)
+		}
+	}
+	return nil
+}
+
+// Verify that a range proof is valid.
+func (proof *KeyRangeProof) Verify(
+	startKey, endKey []byte, limit int, keys, values [][]byte, root []byte,
+) error {
+	if len(proof.PathToKeys) != len(keys) || len(values) != len(keys) {
+		return errors.New("wrong number of keys or values for proof")
+	}
+	if len(proof.PathToKeys) == 0 && proof.LeftPath.isEmpty() && proof.RightPath.isEmpty() {
+		return errors.New("proof is incomplete")
+	}
+
+	// If startKey > endKey, reverse the keys and values, since our proofs are
+	// always in ascending order.
+	if bytes.Compare(startKey, endKey) == 1 {
+		startKey, endKey = endKey, startKey
+
+		ks := make([][]byte, len(keys))
+		vs := make([][]byte, len(keys))
+		for i, _ := range keys {
+			ks[len(ks)-1-i] = keys[i]
+			vs[len(vs)-1-i] = values[i]
+		}
+		keys, values = ks, vs
+	}
+
+	if len(keys) > 0 {
+		if !bytes.Equal(startKey, keys[0]) && proof.LeftPath == nil {
+			if !proof.PathToKeys[0].isLeftmost() {
+				return errors.New("left path is nil and first inner path is not leftmost")
+			}
+		}
+		if !bytes.Equal(endKey, keys[len(keys)-1]) && proof.RightPath == nil {
+			if !proof.PathToKeys[len(proof.PathToKeys)-1].isRightmost() {
+				return errors.New("right path is nil and last inner path is not rightmost")
+			}
+		}
+	}
+
+	// If we've used a limit, our end key can be much greater than the right
+	// node key, ex: with the query start=0, end=FF, limit=10, the returned
+	// key range might be from 12 to D4, and the right node key will be the
+	// one following D4, say DF. In this case, we set our start and end keys
+	// to the key range returned.
+	if len(keys) == limit && limit > 0 {
+		endKey = keys[len(keys)-1]
+		startKey = keys[0]
```

Comment:
Yeah I fixed this in one of the previous comments. The only thing we really need to check is `len(keys) > 0`, in which case we can check `keys[0]` and `keys[-1]`.